ccf_transaction_processor/Readme.md

```diff
@@ -6,37 +6,142 @@ https://creativecommons.org/licenses/by/4.0/
+
+        A. Add the following lines to the CMakelists.txt found at $PDO_SRC/ccf_transaction_processor/CCF.
+           (look for other add_ccf_app in this file, add just above or below this)
+
+                add_ccf_app(
+		   pdoenc SRCS ../transaction_processor/pdo_tp.cpp
+		   ../transaction_processor/verify_signatures.cpp
+		)
+
+        B. Create the build folder
+
+                ```bash
+		mkdir $PDO_SRC/ccf_transaction_processor/CCF/build
+		```
+
+        C. Set the build flags
+
+		```bash
+                cd $PDO_SRC/ccf_transaction_processor/CCF/build
+                cmake -GNinja -DCOMPILE_TARGETS=virtual -DBUILD_END_TO_END_TESTS=OFF -DBUILD_SMALLBANK=OFF \
+                              -DBUILD_TESTS=OFF-DBUILD_UNIT_TESTS=OFF ..
+		```
+
+
+        D. Build using ninja
+
+                ```bash
+		ninja
+		```
+
+           If the VM has a small memory (< 4GB), it might be required to reduce the parallelism in the build process using the `-j` flag for
+           the `ninja` command, say ninja -j2 or even ninja -j1 (the default seems to be ninja -j4).
+
+
+If build is successful, the libpdoenc.virtual.so library is created inside the build folder (ignore any other targets that get created)
+
+
+# Deploy CCF with PDO-TP as the application
+
+The following commands start a single node CCF under the virtual enclave mode.  PDO-TP is the hosted application.
+
+```bash
+cd $PDO_SRC/ccf_transaction_processor/CCF/build
+python3.7 -m venv env
+source env/bin/activate
+pip install -q -U -r ../tests/requirements.txt
+python ../tests/start_network.py --gov-script ../src/runtime_config/gov.lua  --label pdo_tp -e virtual --package libpdoenc.virtual.so --node <ip-address:6006>
+```
+
+A single node CCF instance hosting the PDO TP will be available for business use @ ip-address:6006. Here `ip-address` is the address of this VM that can be used to ping this machine from other VMs (this is our use case). Set ip-address to 127.0.0.1 for local testing. If you are behind a proxy, add ip-address to the list of no_proxy bash env variable.
+
+# Share CCF Keys
+
+CCF uses mutually authenticated TLS channels for transactions. Keys are located at
+$PDO_SRC/ccf_transaction_processor/CCF/build/workspace/pdo_tp_common. The network certificate is `networkcert.pem`.
+User public certificate is `user0_cert.pem` and private key is `user0_privk.pem`.
+
+In our usage, CCF users are PDO clients, and PDO client authentication is implemented within the transaction processor itself. Thus, we do not utilize the client authentication feature provided by CCF. However to satisfy the CCF's requirement that only authorized CCF users can submit transactions to CCF, share `user0_cert.pem` and `user0_privk.pem` with all the PDO clients. These two keys must be stored under $PDO_LEDGER_KEY_ROOT (as part of PDO deployment) under the names `userccf_cert.pem` and `userccf_privk.pem` respectively.
```

Comment:
I would have assumed we also would have to somewhere have a trusted version of the CCFs PK (`networkcert.pem`) to authenticate that we are really talking to the right CCF?  Certainly would seem to be crucial for what you say later about signed Payloads of read transaction below?

Reply:
Actually, you _do_ require that networkcert.pem is also copied to PDO_LEDGER_KEY_ROOT!

Reply:
btw: seems this copying has to be done _each time_ i restart, even if i already have a workspace?  Is there a way to start the TP on existing keys?

Reply:
Yes, thank you, you will have to store the network keys as well. And, I would not imagine that there is a way, since the keys are unique to an instance of the CCF enclave.

Reply:
well, i was wondering about a multiple instances of the nodes but for a single instance of a workspace.  That said, it is true that the whole idea with Raft+SGX really relies on having at any point in time some running enclaves and having all (bootstrap/root) keys/state only memory and never stored/sealed.  Insofar i guess my question didn't really make sense ...